### PR TITLE
Update ScopedViewModel to use viewModelScope

### PR DESCRIPTION
**Pull request by hichamboushaba**
This PR's goal is to update `ScopedViewModel` to use the androidx `viewModelScope`, this is needed to avoid any unwanted hidden bugs caused by our implementation.
`viewModelScope` uses a `SupervisorJob` which allows handling child jobs failures, instead of dying and cancelling the all the jobs, and also uses [`Dispatchers.Main.immediate`](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-main-coroutine-dispatcher/immediate.html), which would bring some performance gains by avoiding re-dispatching to the same dispatcher.

While discussing the matter [before](https://github.com/woocommerce/woocommerce-android/pull/3939#discussion_r623239977), we discussed about the impact on tests, and I thought that it would be small since we can change the `MainDispatcher` in tests, and we already do it using `CoroutineTestRule`, but after trying it, I found out that some tests will require some changes, and the cause is three points:
1. Some tests still uses the function `test` instead of `runBlockingTest`, and don't add the rule `CoroutineTestRule`
2. In some tests that use `runBlockingTest`, the regular function is being used, instead of the extension `TestCoroutineDispatcher.runBlockingTest` which would allow to share the same dispatcher between the ViewModel and the tests.
3. When using `TestCoroutineDispatcher`, delays are skipped automatically, [except](https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-test/README.md#testing-launch-or-async-with-delay) when they are launched from a separate `launch` or `async` builder, which is the case for most of our usage, so we need to explicitly advance time using either `advanceTimeBy` or `advanceUntilIdle`.
IMO, this is a good point, and we should take advantage from it, as it would allow more advanced testing behaviors, like confirming that an action occurs only after the correct time has passed.

### Tips for review:
The PR seems big, but some parts can be ignored while reviewing:
1. I intentionally left all reformatting to a single commit 7e9c2704868fbd95ee03430c9ce9aa901e492ba0, so you can ignore it while reviewing.
2. The commit 40e674d345305cd47016dba011fcc7d96767abc8 is just to move the `CoroutineTestRule` declaration to the BaseUnitTest class, as it's something that we need in all child classes.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/viewmodel/ScopedViewModel.kt [hunk 1]
```diff
@@ -16,22 +15,13 @@ import kotlin.coroutines.CoroutineContext
  * When the ViewModel is destroyed, the coroutine job is cancelled and any running coroutine tied to it is stopped.
  */
 abstract class ScopedViewModel(
-    protected val savedState: SavedStateHandle,
-    protected val dispatchers: CoroutineDispatchers
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/viewmodel/ScopedViewModel.kt:20]: In the same [discussion](https://github.com/woocommerce/woocommerce-android/pull/3939#discussion_r623239977) regarding this subject, it was suggested that we should get rid of the `dispatchers` constructor parameter in the base class, and inject it only in the child viewmodels which need it.
WDYT about this point?

File: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModelTest.kt [hunk 2]
```diff
@@ -247,20 +245,22 @@ class CardReaderPaymentViewModelTest : BaseUnitTest() {
...
+        coroutinesTestRule.testDispatcher.runBlockingTest {
             whenever(cardReaderManager.collectPayment(any(), any(), anyString())).thenAnswer {
                 flow { emit(CardPaymentStatus.CollectingPaymentFailed(mock())) }
             }
             viewModel.start(cardReaderManager)
 
             (viewModel.viewStateData.value as FailedPaymentState).onPrimaryActionClicked.invoke()
 
+            advanceUntilIdle()
+
```
Comment (malinajirka) [inline WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModelTest.kt:261]: 🔍 Nitpick: I usually try to structure unit tests into 3 sections separated by an empty line "Arrange/Act/Assert". I'd consider removing this empty line (or the empty line above the statement. Wdyt?

**Review by malinajirka — COMMENTED** [review]
Thanks @hichamboushaba for working on this! The changes look good to me ;)!

> When using TestCoroutineDispatcher, delays are skipped automatically, except when they are launched from a separate launch or async builder, which is the case for most of our usage, so we need to explicitly advance time using either advanceTimeBy or advanceUntilIdle.
IMO, this is a good point, and we should take advantage from it, as it would allow more advanced testing behaviors, like confirming that an action occurs only after the correct time has passed.

I find both `advanceTimeBy` and `pause/resumeDispatcher` methods quite useful in some scenarios. I just wish the default behavior would advanceUntilIdle automatically and we could set a flag to enable manual handling. I'm a bit worried that we might not realize that we need to manually call "advanceUntilIdle" when writing tests and we might get stuck on it and waste time but I guess we can get used to it.

File: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModelTest.kt (hunk 2 above)
Comment (hichamboushaba) [reply WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModelTest.kt:261]: I agree, removing the above empty line is a good thing to clearly state that this `advancement` is part of the Act section.
I'll update it :)

Thanks for taking a look at the PR 🙏

**Review by malinajirka — APPROVED** [review]
🥇 Thank you for working on this @hichamboushaba! 

LGTM :shipit:

**Pull request merged**